src/OrchardCore/OrchardCore.Media.Core/OrchardCore.Media.Core/DefaultMediaFileStoreCacheFileProvider.cs

```diff
@@ -47,8 +47,9 @@ public Task<bool> IsCachedAsync(string path)
 
         public async Task SetCacheAsync(Stream stream, IFileStoreEntry fileStoreEntry, CancellationToken cancellationToken)
         {
-            // File store semantics include a leading slash.
+            // File store semantics include a leading slash, and may included escaped characters.
             var cachePath = Path.Combine(Root, fileStoreEntry.Path.Substring(1));
```

Comment:
What kind of encoding? Also you are removing the first char of an encoded string, so could the first char actually be encoded and removing the first part of it break everything?

Reply:
A file named `/â, ê, î, ô, û/â, ê, î, ô, û Δ Σ Ψ Ψ 汉字 漢字/Ç à è ù Δ Σ Ψ Ψ 漢字 漢字.jpg`
gets encoded presumably by razor, and arrives encoded in the url path like this (it's PathString so always has a leading /)

`{/%C3%A2,%20%C3%AA,%20%C3%AE,%20%C3%B4,%20%C3%BB/%C3%A2,%20%C3%AA,%20%C3%AE,%20%C3%B4,%20%C3%BB%20%20%20%20%20%20%CE%94%20%CE%A3%20%CE%A8%20%20%20%CE%A8%20%20%20%20%E6%B1%89%E5%AD%97%20%20%20%20%20%E6%BC%A2%E5%AD%97/%C3%87%20%C3%A0%20%C3%A8%20%C3%B9%20%CE%94%20%CE%A3%20%CE%A8%20%CE%A8%20%E6%BC%A2%E5%AD%97%20%20%20%20%20%20%E6%BC%A2%E5%AD%97.jpg}`

Then when they go into blob they go in with that path, it strips the leading slash, then unescapes it. The blob library then escapes it again when it does it's upload, and records the filepath in filestoreEntry.Path with a leading slash (file store semantics), but with the escaped blob path.

So when it saves to disc here, it needs to be unescaped, and have the leading `/` removed, so it can go in the the physical file system, it saves unescaped. Otherwise it saves as `%C3%A2` etc.

The physical file provider also does unescaping, so later on during retrieval of the PathString/url, it'll be able to retrieve essentially the NTFS file name (unescaped).

Reply:
> gets encoded presumably by razor

Then I will debug to understand what actually happens, it might be an issue at this point and not where you fixed it.

Reply:
> but hovering over it gives the url encoded path

Which seems ok, it's the same url but encoded. Which makes me think that we should have encoded it when the link was generated. We probably not do it as often as we should do and only focus on query string segments, where the whole media path should be encoded actually (between each / probably). Even though if the path comes from the filesystem or blob storage the same chars should be valid, and secure (two different things that both matter).

Then the concern is that on the server when you get the request, it should be available as a decoded value, you should not have to decode it yourself. I think there is an issue there. Or that you are reading the wrong property that gets what was actually sent (the encoded value).

Reply:
Yeah I think you're right, this might be fixed in the wrong place.

It depends what we want the value of an `IFileStoreEntry.Path` to be.
Should it be the underlying blob encoded path, or the path that externally we would know it as?

i.e. 
1. 20%20%20%20%E6%BC%A2%E5%AD%97.jpg
2. Ç à è ù Δ Σ Ψ Ψ 漢字 漢字.jpg 

Probably the second, for GraphQL when it queries, should recieve an unescaped path I suspect?

As you debug if you agree with the 2nd option let me know and I'll rearrange where I've unescaped.

Reply:
2nd option obviously

Reply:
I learn something every day

Turns out using PathString directly returns the escaped request path, e.g.
`/media/%C3%A2,%20%C3%AA,%20%C3%AE.jpg`

Using PathString.Value returns the unescaped request path, e.g.
`/media/â, ê, î, ô, û/â, ê, î, ô, û      Δ Σ Ψ   Ψ    汉字     漢字/Ç à è ù Δ Σ Ψ Ψ 漢字      漢字.jpg`

So have adapted to use subPath.Value in hopefully the right place

**Note** While testing GraphQL there is another unrelated issue with `BlogFileStore.GetDirectoryContentAsync` where it doesn't recurse the directories, (there is no code for it to do so from first glance), so I'll open another issue for that, as it seems unrelated.

Reply:
OK tested this with GraphQl (w/ the potential fix for it in) and others and getting the same results back for both local files and blob files.

GraphQl response example
```
      {
        "lastModifiedUtc": "2019-10-03T17:21:20Z",
        "path": "/agencycache/media/Δ Σ Ψ.jpg"
      }
```

So I think this is done in the right place now.

Noting it doesn't address any of the concerns about whether we should be encoding these values in the image / razor / liquid helpers.

Reply:
Yes, e.g every time there is an implicit call to `ToString()` on a `PathString`, e.g when passed to a method that uses a string parameter (not a `PathString`)

        public override string ToString()
        {
            return ToUriComponent();
        }

Hmm, maybe we need to check if there are other places where it would need to be updated.

Reply:
Maybe we need to update all the place that a `PathString` is used as it is (or using explicitly its `ToString()`) to set a string property, string variable, passed as a string parameter and so on.

e.g replace

    if (_entries.TryGetContentItemId(httpContext.Request.Path.ToString().TrimEnd('/'), ...))

With

    if (_entries.TryGetContentItemId(httpContext.Request.Path.Value.TrimEnd('/'), ...))


The good new is that it will fix some potential issues and it will be faster by not calling `ToUriComponent()`.